### ISSUE TITLE
1.1.1-pre2 sparc64 failed poly1305_internal_test

### ISSUE THREAD

**Issue by sebastianas** [openssl/openssl#5586]
The build log says:
```
../../test/recipes/03-test_internal_poly1305.t ........
1..1
    # Subtest: ../../test/poly1305_internal_test
    1..1
        # Subtest: test_poly1305
        1..35
../../util/shlib_wrap.sh ../../test/poly1305_internal_test => 139
not ok 1 - running poly1305_internal_test

#   Failed test 'running poly1305_internal_test'
#   at /home/bigeasy/openssl-1.1.1~~pre2/buil2/test/../../util/perl/OpenSSL/Test/Simple.pm line 77.
# Looks like you failed 1 test of 1.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests
```
and invoking test/poly1305_internal_test in gdb gave me:

```
1..1
    # Subtest: test_poly1305
    1..35

Program received signal SIGSEGV, Segmentation fault.
0x000001000009a400 in poly1305_init ()
(gdb) bt
#0  0x000001000009a400 in poly1305_init ()
#1  0x000001000005c470 in Poly1305_Init ()
#2  0x00000100000393b4 in test_poly1305 ()
#3  0x0000010000039dc4 in run_tests ()
#4  0x000001000003913c in main ()
(gdb) disassemble
Dump of assembler code for function poly1305_init:
   0x000001000009a3e0 <+0>:     save  %sp, -208, %sp
   0x000001000009a3e4 <+4>:     nop
   0x000001000009a3e8 <+8>:     sethi  %hi(0x289c00), %g1
   0x000001000009a3ec <+12>:    call  0x1000009a3c0
   0x000001000009a3f0 <+16>:    or  %g1, 0x14, %g1      ! 0x289c14
   0x000001000009a3f4 <+20>:    sethi  %hi(0x800), %o7
   0x000001000009a3f8 <+24>:    or  %o7, 0x108, %o7     ! 0x908
   0x000001000009a3fc <+28>:    ldx  [ %g1 + %o7 ], %g1
=> 0x000001000009a400 <+32>:    ld  [ %g1 ], %g1
   0x000001000009a404 <+36>:    and  %g1, 0x50, %g1
   0x000001000009a408 <+40>:    cmp  %g1, 0x10
   0x000001000009a40c <+44>:    be  0x1000009a868 <poly1305_init_fma+8>
   0x000001000009a410 <+48>:    nop
   0x000001000009a414 <+52>:    clrx  [ %i0 ]
   0x000001000009a418 <+56>:    clrx  [ %i0 + 8 ]

```
It was configured as "../Configure linux64-sparcv9" on UltraSparc T5 (Niagara5) running Debian unstable.

**Comment by dot-asm** [comment]
For the record. I perform periodic tests of SPARC Linux and haven't encountered anything of the sort. Nor can I reproduce it right now. Anyway, what's that instruction about? It's all about position-independent code. That thing with call 0x1000009a3c0 few lines up is supposed to compose a reference to _GLOBAL_OFFSET_TABLE_, then it composes offset to slot that holds OPENSSL_sparcv9cap_P's address, pulls the address and finally attempts to dereference it. So the crash means that either pointer to OPENSSL_sparcv9cap_P in offset table is corrupted, or [more likely] constants are off and code is picking some random data and treats it as a pointer. Note that assembler is responsible for generating relocations in object file, and linker is responsible for filling the actual values. This brings us to following question. What is your binutils' version? I have no problems with 2.29.1. Submit even output from 'objdump -r crypto/poly1305/poly1305-sparcv9.o'...

**Comment by dot-asm** [comment]
There is one subtle thing that changed recently, so try following. `rm test/poly1305_internal_test.o`, then `make -n test/poly1305_internal_test.o`. Next copy-n-paste the command, *add* -fPIC flag, and execute it. Finally `make test/poly1305_internal_test` and `test/poly1305_internal_test`. Does it crash even then?

**Comment by sebastianas** [comment]
```
objdump -r crypto/poly1305/poly1305-sparcv9.o                                                                         

crypto/poly1305/poly1305-sparcv9.o:     file format elf64-sparc

RELOCATION RECORDS FOR [.text]:
OFFSET           TYPE              VALUE 
0000000000000028 R_SPARC_PC22      _GLOBAL_OFFSET_TABLE_-0x0000000000000004
0000000000000030 R_SPARC_PC10      _GLOBAL_OFFSET_TABLE_+0x0000000000000004
0000000000000034 R_SPARC_GOT22     OPENSSL_sparcv9cap_P
0000000000000038 R_SPARC_GOT10     OPENSSL_sparcv9cap_P

```
re-compile with -fPIC or -FPIE did also crash. I have here
  GNU objdump (GNU Binutils for Debian) 2.30
and 
```
$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/sparc64-linux-gnu/7/lto-wrapper
Target: sparc64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 7.3.0-10' --with-bugurl=file:///usr/share/doc/gcc-7/README.Bugs --enable-languages=c,ada,c++,go,d,fortran,objc,obj-c++ --prefix=/usr --with-gcc-major-version-only --program-suffix=-7 --program-prefix=sparc64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-libquadmath --disable-libquadmath-support --enable-plugin --enable-default-pie --with-system-zlib --enable-objc-gc=auto --enable-multiarch --disable-werror --with-cpu-32=ultrasparc --enable-targets=all --with-long-double-128 --enable-multilib --enable-checking=release --build=sparc64-linux-gnu --host=sparc64-linux-gnu --target=sparc64-linux-gnu
Thread model: posix
gcc version 7.3.0 (Debian 7.3.0-10)
```
note it has --enable-default-pie. At the time of crash g1 is 0x0 if it is of any help. One additional thing. Decompile part of the code:

```
   9a780:       81 c3 e0 08     retl 
   9a784:       82 03 c0 01     add  %o7, %g1, %g1
   9a788:       30 68 00 06     b,a   %xcc, 9a7a0 <poly1305_init>
   9a78c:       01 00 00 00     nop 
   9a790:       01 00 00 00     nop 
   9a794:       01 00 00 00     nop 
   9a798:       01 00 00 00     nop 
   9a79c:       01 00 00 00     nop 

000000000009a7a0 <poly1305_init>:
   9a7a0:       9d e3 bf 30     save  %sp, -208, %sp
   9a7a4:       01 00 00 00     nop 
   9a7a8:       03 00 0a 6e     sethi  %hi(0x29b800), %g1
   9a7ac:       7f ff ff f5     call  9a780 <ERR_load_PKCS7_strings+0x60>
   9a7b0:       82 10 60 54     or  %g1, 0x54, %g1      ! 29b854 <__FRAME_END__+0x8d1e4>
   9a7b4:       1f 00 00 02     sethi  %hi(0x800), %o7

```
And this is what
```
long asd;

long ret_asd(void)
{
        return asd;
}

```
looks compiled (-c -O2):

```
Disassembly of section .text:

0000000000000000 <ret_asd>:
   0:   9d e3 bf 50     save  %sp, -176, %sp
   4:   2f 00 00 00     sethi  %hi(0), %l7
                        4: R_SPARC_PC22 _GLOBAL_OFFSET_TABLE_-0x4
   8:   40 00 00 00     call  8 <ret_asd+0x8>
                        8: R_SPARC_WPLT30       __sparc_get_pc_thunk.l7
   c:   ae 05 e0 00     add  %l7, 0, %l7        ! 0 <ret_asd>
                        c: R_SPARC_PC10 _GLOBAL_OFFSET_TABLE_+0x4
  10:   03 00 00 00     sethi  %hi(0), %g1
                        10: R_SPARC_GOTDATA_OP_HIX22    asd
  14:   82 18 60 00     xor  %g1, 0, %g1
                        14: R_SPARC_GOTDATA_OP_LOX10    asd
  18:   c2 5d c0 01     ldx  [ %l7 + %g1 ], %g1
                        18: R_SPARC_GOTDATA_OP  asd
  1c:   81 cf e0 08     return  %i7 + 8
  20:   d0 58 40 00     ldx  [ %g1 ], %o0

Disassembly of section .text.startup:

0000000000000000 <main>:
   0:   81 c3 e0 08     retl 
   4:   90 10 20 00     clr  %o0

Disassembly of section .text.__sparc_get_pc_thunk.l7:

0000000000000000 <__sparc_get_pc_thunk.l7>:
   0:   81 c3 e0 08     retl 
   4:   ae 03 c0 17     add  %o7, %l7, %l7
```

```
$ objdump -r ~/a.o 

/home/bigeasy/a.o:     file format elf64-sparc

RELOCATION RECORDS FOR [.text]:
OFFSET           TYPE              VALUE 
0000000000000004 R_SPARC_PC22      _GLOBAL_OFFSET_TABLE_-0x0000000000000004
0000000000000008 R_SPARC_WPLT30    __sparc_get_pc_thunk.l7
000000000000000c R_SPARC_PC10      _GLOBAL_OFFSET_TABLE_+0x0000000000000004
0000000000000010 R_SPARC_GOTDATA_OP_HIX22  asd
0000000000000014 R_SPARC_GOTDATA_OP_LOX10  asd
0000000000000018 R_SPARC_GOTDATA_OP  asd

```

**Comment by dot-asm** [comment]
> objdump -r crypto/poly1305/poly1305-sparcv9.o                                                                         

Looks identical to mine.

> I have here GNU objdump (GNU Binutils for Debian) 2.30.

Can you test on system with previous binutils version? Another question what are your linker flags? Well, if they are different from our defaults. Maybe I can reproduce it...

**Comment by sebastianas** [comment]
On 2018-03-11 18:32:50 [+0000], Andy Polyakov wrote:
> Looks identical to mine.
> 
> > I have here GNU objdump (GNU Binutils for Debian) 2.30.
> 
> Can you test on system with previous binutils version? 
This is a porter box. I can't downgrade.

> Another question what are your linker flags? Well, if they are
> different from our defaults. Maybe I can reproduce it...

Is this helping:

|Command line (with current working directory = .):
|
|    /usr/bin/perl ../Configure linux64-sparcv9
|
|Perl information:
|
|    /usr/bin/perl
|    5.26.1 for sparc64-linux-gnu-thread-multi
|
|Enabled features:
|
|    afalgeng
|    aria
|    asm
|    async
|    autoalginit
|    autoerrinit
|    bf
|    blake2
|    camellia
|    capieng
|    cast
|    chacha
|    cmac
|    cms
|    comp
|    ct
|    deprecated
|    des
|    dgram
|    dh
|    dsa
|    dso
|    dtls
|    dynamic-engine
|    ec
|    ec2m
|    ecdh
|    ecdsa
|    engine
|    err
|    filenames
|    gost
|    hw(-.+)?
|    idea
|    makedepend
|    md4
|    mdc2
|    multiblock
|    nextprotoneg
|    ocb
|    ocsp
|    pic
|    poly1305
|    posix-io
|    psk
|    rc2
|    rc4
|    rdrand
|    rfc3779
|    rmd160
|    scrypt
|    seed
|    shared
|    siphash
|    sm3
|    sm4
|    sock
|    srp
|    srtp
|    sse2
|    ssl
|    static-engine
|    stdio
|    tests
|    threads
|    tls
|    ts
|    ui-console
|    whirlpool
|    tls1
|    tls1-method
|    tls1_1
|    tls1_1-method
|    tls1_2
|    tls1_2-method
|    tls1_3
|    dtls1
|    dtls1-method
|    dtls1_2
|    dtls1_2-method
|
|Disabled features:
|
|    asan                    [default] OPENSSL_NO_ASAN
|    crypto-mdebug           [default] OPENSSL_NO_CRYPTO_MDEBUG
|    crypto-mdebug-backtrace [default] OPENSSL_NO_CRYPTO_MDEBUG_BACKTRACE
|    devcryptoeng            [default] OPENSSL_NO_DEVCRYPTOENG
|    ec_nistp_64_gcc_128     [default] OPENSSL_NO_EC_NISTP_64_GCC_128
|    egd                     [default] OPENSSL_NO_EGD
|    external-tests          [default] OPENSSL_NO_EXTERNAL_TESTS
|    fuzz-libfuzzer          [default] OPENSSL_NO_FUZZ_LIBFUZZER
|    fuzz-afl                [default] OPENSSL_NO_FUZZ_AFL
|    heartbeats              [default] OPENSSL_NO_HEARTBEATS
|    md2                     [default] OPENSSL_NO_MD2 (skip crypto/md2)
|    msan                    [default] OPENSSL_NO_MSAN
|    rc5                     [default] OPENSSL_NO_RC5 (skip crypto/rc5)
|    sctp                    [default] OPENSSL_NO_SCTP
|    ssl-trace               [default] OPENSSL_NO_SSL_TRACE
|    tls13downgrade          [default] OPENSSL_NO_TLS13DOWNGRADE
|    ubsan                   [default] OPENSSL_NO_UBSAN
|    unit-test               [default] OPENSSL_NO_UNIT_TEST
|    weak-ssl-ciphers        [default] OPENSSL_NO_WEAK_SSL_CIPHERS
|    zlib                    [default] 
|    zlib-dynamic            [default] 
|    ssl3                    [default] OPENSSL_NO_SSL3
|    ssl3-method             [default] OPENSSL_NO_SSL3_METHOD
|
|Config target attributes:
|
|    AR => "ar",
|    ARFLAGS => "r",
|    CC => "gcc",
|    CFLAGS => "-Wall -O3",
|    CXX => "g++",
|    CXXFLAGS => "-Wall -O3",
|    HASHBANGPERL => "/usr/bin/env perl",
|    RANLIB => "ranlib",
|    RC => "windres",
|    aes_asm_src => "aes_core.c aes_cbc.c aes-sparcv9.S aest4-sparcv9.S aesfx-sparcv9.S",
|    aes_obj => "aes_core.o aes_cbc.o aes-sparcv9.o aest4-sparcv9.o aesfx-sparcv9.o",
|    apps_aux_src => "",
|    apps_init_src => "",
|    apps_obj => "",
|    bf_asm_src => "bf_enc.c",
|    bf_obj => "bf_enc.o",
|    bn_asm_src => "asm/sparcv8plus.S sparcv9-mont.S sparcv9a-mont.S vis3-mont.S sparct4-mont.S sparcv9-gf2m.S",
|    bn_obj => "asm/sparcv8plus.o sparcv9-mont.o sparcv9a-mont.o vis3-mont.o sparct4-mont.o sparcv9-gf2m.o",
|    bn_ops => "BN_LLONG RC4_CHAR",
|    build_file => "Makefile",
|    build_scheme => [ "unified", "unix" ],
|    cast_asm_src => "c_enc.c",
|    cast_obj => "c_enc.o",
|    cflags => "-pthread -m64 -mcpu=ultrasparc",
|    chacha_asm_src => "chacha_enc.c",
|    chacha_obj => "chacha_enc.o",
|    cmll_asm_src => "camellia.c cmll_misc.c cmll_cbc.c cmllt4-sparcv9.S",
|    cmll_obj => "camellia.o cmll_misc.o cmll_cbc.o cmllt4-sparcv9.o",
|    cppflags => "",
|    cpuid_asm_src => "sparcv9cap.c sparccpuid.S",
|    cpuid_obj => "sparcv9cap.o sparccpuid.o",
|    cxxflags => "-std=c++11 -pthread -m64 -mcpu=ultrasparc",
|    defines => [  ],
|    des_asm_src => "des_enc-sparc.S fcrypt_b.c dest4-sparcv9.S",
|    des_obj => "des_enc-sparc.o fcrypt_b.o dest4-sparcv9.o",
|    disable => [  ],
|    dso_extension => ".so",
|    dso_scheme => "DLFCN",
|    ec_asm_src => "ecp_nistz256.c ecp_nistz256-sparcv9.S",
|    ec_obj => "ecp_nistz256.o ecp_nistz256-sparcv9.o",
|    enable => [ "afalgeng" ],
|    ex_libs => "-ldl -pthread",
|    exe_extension => "",
|    includes => [  ],
|    lflags => "",
|    lib_cflags => "",
|    lib_cppflags => "-DOPENSSL_USE_NODELETE -DB_ENDIAN",
|    lib_defines => [  ],
|    md5_asm_src => "md5-sparcv9.S",
|    md5_obj => "md5-sparcv9.o",
|    modes_asm_src => "ghash-sparcv9.S",
|    modes_obj => "ghash-sparcv9.o",
|    module_cflags => "-fPIC",
|    module_cxxflags => "",
|    module_ldflags => "-Wl,-znodelete -shared",
|    multilib => "64",
|    padlock_asm_src => "",
|    padlock_obj => "",
|    perlasm_scheme => "void",
|    poly1305_asm_src => "poly1305-sparcv9.S",
|    poly1305_obj => "poly1305-sparcv9.o",
|    rc4_asm_src => "rc4_enc.c rc4_skey.c",
|    rc4_obj => "rc4_enc.o rc4_skey.o",
|    rc5_asm_src => "rc5_enc.c",
|    rc5_obj => "rc5_enc.o",
|    rmd160_asm_src => "",
|    rmd160_obj => "",
|    sha1_asm_src => "sha1-sparcv9.S sha256-sparcv9.S sha512-sparcv9.S",
|    sha1_obj => "sha1-sparcv9.o sha256-sparcv9.o sha512-sparcv9.o",
|    shared_cflag => "-fPIC",
|    shared_defflag => "-Wl,--version-script=",
|    shared_defines => [  ],
|    shared_extension => ".so.\$(SHLIB_VERSION_NUMBER)",
|    shared_extension_simple => ".so",
|    shared_ldflag => "-Wl,-znodelete -shared",
|    shared_rcflag => "",
|    shared_sonameflag => "-Wl,-soname=",
|    shared_target => "linux-shared",
|    thread_defines => [  ],
|    thread_scheme => "pthreads",
|    unistd => "<unistd.h>",
|    uplink_aux_src => "",
|    uplink_obj => "",
|    wp_asm_src => "wp_block.c",
|    wp_obj => "wp_block.o",
|
|Recorded environment:
|
|    AR = 
|    ARFLAGS = 
|    AS = 
|    ASFLAGS = 
|    BUILDFILE = 
|    CC = 
|    CFLAGS = 
|    CPP = 
|    CPPDEFINES = 
|    CPPFLAGS = 
|    CPPINCLUDES = 
|    CROSS_COMPILE = 
|    CXX = 
|    CXXFLAGS = 
|    HASHBANGPERL = 
|    LD = 
|    LDFLAGS = 
|    LDLIBS = 
|    MT = 
|    MTFLAGS = 
|    OPENSSL_LOCAL_CONFIG_DIR = 
|    PERL = 
|    RANLIB = 
|    RC = 
|    RCFLAGS = 
|    RM = 
|    WINDRES = 
|    __CNF_CFLAGS = 
|    __CNF_CPPDEFINES = 
|    __CNF_CPPFLAGS = 
|    __CNF_CPPINCLUDES = 
|    __CNF_CXXFLAGS = 
|    __CNF_LDFLAGS = 
|    __CNF_LDLIBS = 
|
|Makevars:
|
|    AR              = ar
|    ARFLAGS         = r
|    CC              = gcc
|    CFLAGS          = -Wall -O3
|    CPPDEFINES      = 
|    CPPFLAGS        = 
|    CPPINCLUDES     = 
|    CXX             = g++
|    CXXFLAGS        = -Wall -O3
|    HASHBANGPERL    = /usr/bin/env perl
|    LDFLAGS         = 
|    LDLIBS          = 
|    RANLIB          = ranlib
|    RC              = windres
|
|NOTE: These variables only represent the configuration view.  The build file
|template may have processed these variables further, please have a look at the
|build file for more exact data:
|    Makefile
|
|build file:
|
|    Makefile
|
|build file templates:
|
|    ../Configurations/unix-Makefile.tmpl
|    ../Configurations/common.tmpl

Or do you need something different?

Sebastian

**Comment by dot-asm** [comment]
> Or do you need something different?

Hmm, I don't see -Bsymbolic in shared_ldflags. So it's not default flags... Could you add it back to see if it works. Though OPENSSL_sparcv9cap_P is marked with __attribute__ ((visibility("hidden"))), so it should be "wired" as if it was -Bsymbolic...

**Comment by andwat18** [comment]
pycache__/
_build/
build/
dist/
htmlcov/
*.so
.tox/
.cache/
.coverage
*.egg-info/
*.egg
.eggs/
*.py[co]
.hypothesis/

**Comment by sebastianas** [comment]
On 2018-03-11 19:01:46 [+0000], Andy Polyakov wrote:
> > Or do you need something different?
> 
> Hmm, I don't see -Bsymbolic in shared_ldflags. So it's not default
> flags... Could you add it back to see if it works. Though
> OPENSSL_sparcv9cap_P is marked with __attribute__
> ((visibility("hidden"))), so it should be "wired" as if it was
> -Bsymbolic...

I don't recall why we drop that, @kroeckx should know. Anyway, I added
it back and it is still crashing.
poly1305-sparcv9 is the only user of SPARC_LOAD_ADDRESS. There are other
using SPARC_LOAD_ADDRESS_LEAF with that variable. Are those user covered
by the testuite? Because if they are, they are not failing.

Sebastian

**Comment by dot-asm** [comment]
All other cases are tested by calling into libcrypto.so where OPENSSL_sparcv9cap_P is linked by the time libcrypto.so was linked. And that apparently works. I mean I read your report that poly1305_internal_test is the *only* one failing. What's special about poly1305_internal_test is that it's linked with libcrypto.a instead. I don't think it has anything to do with LEAF vs. no-LEAF. Can you test with previous binutils version? I'm not suggesting it as permanent solution (unless it's proven to be a bug), only as essential trouble-shooting step.

**Comment by glaubitz** [comment]
> For the record. I perform periodic tests of SPARC Linux and haven't encountered anything of the sort.

On which particular Linux? Gentoo for SPARC? Oracle Linux for SPARC? Linux for SPARC?

Please note that Debian's SPARC uses a 64-bit userland while Gentoo Linux is based on 32-bit SPARCv9.

CC @jrtc27

**Comment by dot-asm** [comment]
My remark about -Bsymbolic was irrelevant, because, as I said myself, poly1305_internal_test is linked with libcrypto.a, and how libcrypto.so is linked has no bearing. I've tried to link it with ld 2.30 and still can't reproduce the problem. I have no control of the system I have access to, so it's not like I upgraded binutils package, I can't do that. So I've compiled binutils copied ld to /some/where and passed -B/some/where to compiler...

**Comment by dot-asm** [comment]
What does `nm test/poly1305_internal_test | grep sparcv9cap` say? If I set break point to poly1305_init and single step, I get same value in %g1 as in nm output.

**Comment by sebastianas** [comment]
On 2018-03-12 03:21:40 [-0700], Andy Polyakov wrote:
> My remark about -Bsymbolic was irrelevant, because, as I said myself,
> poly1305_internal_test is linked with libcrypto.a, and how
> libcrypto.so is linked has no bearing. I've tried to link it with ld
> 2.30 and still can't reproduce the problem. I have no control of the
> system I have access to, so it's not like I upgraded binutils package,
> I can't do that. So I've compiled binutils copied ld to /some/where
> and passed -B/some/where to compiler...

This is what I did, too. binutils 2.30-4 works, 2.30-5 does not. I would
blame the patch [0] which was included as part of "branch-updates.diff".
Maybe something else but I identified a good and a bad one. Could a
sparc porter please take a lookg this :)

[0] https://bugs.debian.org/cgi-bin/bugreport.cgi?att=1;bug=890359;filename=sparc64-fix-got.patch;msg=5

Sebastian

**Comment by sebastianas** [comment]
On 2018-03-12 07:40:48 [-0700], Andy Polyakov wrote:
> What does `nm test/poly1305_internal_test | grep sparcv9cap` say? If I set break point to poly1305_init and single step, I get same value in %g1 as in nm output.

nm test/poly1305_internal_test | grep sparcv9cap
00000000003380f0 D OPENSSL_sparcv9cap_P

At the point of crash %g1 is NULL.

Sebastian

**Comment by jrtc27** [comment]
This is a binutils bug filed as https://sourceware.org/bugzilla/show_bug.cgi?id=22972.

The TL;DR is that poly1305-sparcv9.pl (.S) uses the old GOT2210 relocations rather than the newer GOTDATA_OP_HIX22/LOX10 (and GOTDATA_OP on the load) relocations, where the latter can be transformed by the linker to a GOT-relative add if the symbol known to be within the same object, which also removes the need for the GOT entry. Unfortunately if ld sees the GOTDATA_OP relocations first, it immediately decides to do the optimisation before it's seen the GOT22/10 relocations, which is of course wrong when there are GOT22/10 relocations that it will see in subsequent files (well, it can do the optimisation, but must keep the GOT slot as potentially in-use until the end).

**Comment by dot-asm** [comment]
So this means that this issue can be closed. Thanks! On closing note I'd like to remind that OpenSSL doesn't "live" in the world of latest compiler or even specific OS version. What I'm trying to say is that it would be surely possible to argue that module in question should use new fancy relocations, but that would mean that we would have to detect if they are available here and now and act accordingly, i.e. old relocations wouldn't go anywhere. Which in turn naturally means that we would rather not, as we would rather prefer to have kind of a common denominator that works in all cases. Thanks again!
